proposal.md

```diff
@@ -0,0 +1,1451 @@
+providing greater flexibility in usage.
+
+The removal of the `prefix` and `indent` argument from `Indent` improves
+the ergonomics of the method as most users just want indented output
+without thinking about the particular `indent` string used.
+These can still be specified using the `WithIndentPrefix` and `WithIndent` options.
+
+One major criticism of `Canonicalize` (per [RFC 8785](https://www.rfc-editor.org/rfc/rfc8785))
+is that it mangles the precision of wide integers. By accepting options, users can
+additionally specify `CanonicalizeRawInts(false)` to prevent this behavior,
+while still having canonicalization for all other JSON artifacts.
+
+The `Format` method was newly added as the primary implementation backing
+`Compact`, `Indent`, and `Canonicalize`.
+
+The following options were added to provide greater flexibility to formatting:
+```diff
++ func CanonicalizeRawFloats(v bool) Options
++ func CanonicalizeRawInts(v bool) Options
++ func PreserveRawStrings(v bool) Options
++ func ReorderRawObjects(v bool) Options
++ func SpaceAfterComma(v bool) Options
++ func SpaceAfterColon(v bool) Options
+
+- func Expand(v bool) Options
++ func Multiline(v bool) Options
+```
+The `Expand` option was renamed as `Multiline` to be more clear and
+to distinguish it from `SpaceAfterComma` and `SpaceAfterColon`
+(which both technically "expand" the output).
+
+The following formatting API has been added:
+```diff
++ func AppendFormat(dst, src []byte, opts ...Options) ([]byte, error)
++ func AppendQuote[Bytes ~[]byte | ~string](dst []byte, src Bytes) ([]byte, error)
++ func AppendUnquote[Bytes ~[]byte | ~string](dst []byte, src Bytes) ([]byte, error)
+```
+
+The length returned by `StackIndex` is now a `int64` instead of `int`
+since the length of a JSON array or object could theoretically exceed `int`
+when handling JSON in a purely streaming manner.
+The stack depth, however, remains fundamentally limited by
+the amount of system memory, so an `int` is still an appropriate type.
+```diff
+-func (e *Encoder) StackIndex(i int) (Kind, int)
++func (e *Encoder) StackIndex(i int) (Kind, int64)
+-func (d *Decoder) StackIndex(i int) (Kind, int)
++func (d *Decoder) StackIndex(i int) (Kind, int64)
+```
+
+The pointer returned by `StackPointer` is now the named `Pointer` type.
+```diff
+- func (e *Encoder) StackPointer() string
++ func (e *Encoder) StackPointer() Pointer
+- func (d *Decoder) StackPointer() string
++ func (d *Decoder) StackPointer() Pointer
+```
+
+An explicit `Pointer` type was added to represent a JSON Pointer
+([RFC 6901](https://www.rfc-editor.org/rfc/rfc6901))
+as a means to identify exactly where an error occurred.
+Convenience methods are defined for interacting with a pointer.
+```diff
++ type Pointer string
++ func (p Pointer) IsValid() bool
++ func (p Pointer) AppendToken(tok string) Pointer
++ func (p Pointer) Parent() Pointer
++ func (p1 Pointer) Contains(p2 Pointer) bool
```

Comment:
interesting that we went with an inconsistent receiver name here. I'd find `func (p Pointer) Contains (p2 Pointer) bool` to be perfectly fine.